### PR TITLE
[modules][av][Android] Fix libraries using ViewDefinitionBuilder.kt crashes in release with proguard enabled

### LINKED ISSUE #20122: [expo-modules-core][android][SDK 47] Libraries using ViewDefinitionBuilder.kt crashes in release with proguard enabled
### Summary

I discovered this issue when using `expo-camera@13.0.0` with `expo-modules-core@0.13.4 ~ 1.0.2`. Reverting to `expo-camera@12.3.0` fixed the issue. After doing some digging, I found that disabling proguard resolves this issue.

`proguard-rules.pro`

```
# expo-camera
-keep class expo.modules.camera.** {*;}
```

OR

`app/build.gradle`

```gradle
def enableProguardInReleaseBuilds = false
```

I can only guess that since reflection is being used in `ViewDefinitionBuilder.kt`, we have to specifically disable proguard optimisation in files that use this new way of constructing views.

Related PR that may have caused this issue: #18820

### Stack trace

```
 java.lang.IllegalArgumentException: Required value was null.
     at expo.modules.kotlin.views.e$a.a(ViewDefinitionBuilder.kt:16)
     at expo.modules.kotlin.views.e$a.invoke(Unknown Source:4)
     at expo.modules.kotlin.views.g.a(Unknown Source:12)
     at expo.modules.kotlin.views.i.b(ViewManagerWrapperDelegate.kt:2)
     at expo.modules.kotlin.views.GroupViewManagerWrapper.createViewInstance(GroupViewManagerWrapper.kt:2)
     at expo.modules.kotlin.views.GroupViewManagerWrapper.createViewInstance(GroupViewManagerWrapper.kt:1)
     at com.facebook.react.uimanager.ViewManager.createViewInstance(ViewManager.java:4)
     at com.facebook.react.uimanager.ViewManager.createView(ViewManager.java:1)
     at com.facebook.react.uimanager.o.createView(NativeViewHierarchyManager.java:8)
     at com.facebook.react.uimanager.b1$e.g(UIViewOperationQueue.java:2)
     at com.facebook.react.uimanager.b1$a.run(UIViewOperationQueue.java:16)
     at com.facebook.react.uimanager.b1.T(UIViewOperationQueue.java:10)
     at com.facebook.react.uimanager.b1.s(Unknown Source:0)
     at com.facebook.react.uimanager.b1$j.doFrameGuarded(UIViewOperationQueue.java:6)
     at com.facebook.react.uimanager.f.doFrame(GuardedFrameCallback.java:1)
     at com.facebook.react.modules.core.g$d.doFrame(ReactChoreographer.java:7)
     at com.facebook.react.modules.core.a$a$a.doFrame(Unknown Source:2)
     at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1229)
     at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1239)
     at android.view.Choreographer.doCallbacks(Choreographer.java:899)
     at android.view.Choreographer.doFrame(Choreographer.java:827)
     at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1214)
     at android.os.Handler.handleCallback(Handler.java:942)
     at android.os.Handler.dispatchMessage(Handler.java:99)
     at android.os.Looper.loopOnce(Looper.java:201)
     at android.os.Looper.loop(Looper.java:288)
     at android.app.ActivityThread.main(ActivityThread.java:7898)
     at java.lang.reflect.Method.invoke(Native Method)
     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

### What platform(s) does this occur on?

Android

### Environment

```
expo-env-info 1.0.5 environment info:
  System:
    OS: macOS 13.0.1
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.17.1 - ~/.nvm/versions/node/v16.17.1/bin/node
    Yarn: 1.22.19 - ~/Documents/GitHub/CAMS_App/node_modules/.bin/yarn
    npm: 8.15.0 - ~/.nvm/versions/node/v16.17.1/bin/npm
    Watchman: 2022.10.31.00 - /opt/homebrew/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.1, iOS 16.1, macOS 13.0, tvOS 16.1, watchOS 9.1
  IDEs:
    Android Studio: Flamingo 2022.2.1 Canary 8 Flamingo 2022.2.1 Canary 8
    Xcode: 14.1/14B47b - /usr/bin/xcodebuild
  npmPackages:
    expo: ^47.0.6 => 47.0.6 
    react: 18.2.0 => 18.2.0 
    react-native: 0.70.6 => 0.70.6 
  Expo Workflow: bare
```

### Works in release

- `expo-camera@12.3.0` or lower as it is using the old way of creating views

### Not working in release

- `expo-camera@1.0.0` or higher as it is using the new way of creating views with `ViewDefinitionBuilder.kt` in `expo-modules-core`

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Fixes https://github.com/expo/expo/issues/20122.

# How

We're missing a couple of proguard rules to prevent view callbacks and contractors from being optimized by the R8. 
I've added rules that prevent two possible contractors in ExpoView:
- public <init>(android.content.Context);
- public <init>(android.content.Context, expo.modules.kotlin.AppContext);

and properties with types connected to view callbacks:
- expo.modules.kotlin.viewevent.ViewEventCallback
- expo.modules.kotlin.viewevent.ViewEventDelegate

# Test Plan

- bare-expo with NCL built in the release mode ✅

### CODE REVIEW

File: packages/expo-av/CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 ### 🐛 Bug fixes
 
 - Fixed build errors when testing on React Native nightly builds. ([#19805](https://github.com/expo/expo/pull/19805) by [@kudo](https://github.com/kudo))
+- Fixed crashes in the release when the proguard is enabled on Android.
```
Comment (Kudo) [inline packages/expo-av/CHANGELOG.md:12]:
```suggestion
- Fixed crashes when ProGuard or R8 is enabled on Android.
```

File: packages/expo-modules-core/CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,7 @@
 - Fixed build errors when testing on React Native nightly builds. ([#19805](https://github.com/expo/expo/pull/19805) by [@kudo](https://github.com/kudo))
 - Fixed failed resolution of 'java.nio.file.Path' on Android. ([#20037](https://github.com/expo/expo/pull/20037) by [@lukmccall](https://github.com/lukmccall))
 - Fixed views are not correctly initialized after reloading on Android. ([#20063](https://github.com/expo/expo/pull/20063) by [@lukmccall](https://github.com/lukmccall))
+- Fixed libraries using the `ViewDefinitionBuilder` crashes in the release when the proguard is enabled on Android.
```
Comment (Kudo) [inline packages/expo-modules-core/CHANGELOG.md:16]:
```suggestion
- Fixed libraries using the `ViewDefinitionBuilder` crashes when ProGuard or R8 is enabled on Android.
```

**Review by Kudo — APPROVED** [review]
just nit comments for changelogs 😅

**Comment by gloredo** [comment]
Is it already bundleded in `13.0.2`? In the changelog it says Unpublished. It still crashes for me on SDK 47 (managed) when Proguard is enabled, I get `IllegalArgumentException expo.modules.kotlin.views.d$a in a. The required value was null.` from Sentry. Worked with SDK 46 and expo-av `12.0.4`. I think it's related.

**Pull request merged**